retries/retries.go

```diff
@@ -88,55 +95,122 @@ func (et *ErrTimedOut) Unwrap() error {
 	return et.err
 }
 
-func Wait(pctx context.Context, timeout time.Duration, fn WaitFn) error {
-	ctx, cancel := context.WithTimeout(pctx, timeout)
-	defer cancel()
-	var attempt int
-	var lastErr error
-	for {
-		attempt++
-		res := fn()
-		if res == nil {
-			return nil
-		}
-		if res.Halt {
-			return res.Err
-		}
-		lastErr = res.Err
-		wait := Backoff(attempt)
-		timer := time.NewTimer(wait)
-		logger.Tracef(ctx, "%s. Sleeping %s",
-			strings.TrimSuffix(res.Err.Error(), "."),
-			wait.Round(time.Millisecond))
-		select {
-		// stop when either this or parent context times out
-		case <-ctx.Done():
-			timer.Stop()
-			return &ErrTimedOut{lastErr}
-		case <-timer.C:
+// RetryOption is a function that sets part of the retry configuration for a retrier.
+type RetryOption func(*RetryConfig)
+
+// RetryConfig is the configuration for a retrier.
+type RetryConfig struct {
+	timeout     time.Duration
+	shouldRetry func(error) bool
+	backoff     func(int) time.Duration
+}
+
+func (r RetryConfig) Timeout() time.Duration {
+	if r.timeout == 0 {
+		return 20 * time.Minute
+	}
+	return r.timeout
+}
+
+func (r RetryConfig) Backoff(attempt int) time.Duration {
+	if r.backoff == nil {
+		return backoff(attempt)
+	}
+	return r.backoff(attempt)
+}
+
+func (r RetryConfig) ShouldRetry(err error) bool {
+	if r.shouldRetry == nil {
+		return err != nil
+	}
+	return r.shouldRetry(err)
+}
+
+// WithTimeout sets the timeout for the retrier.
+func WithTimeout(timeout time.Duration) RetryOption {
+	return func(rc *RetryConfig) {
+		rc.timeout = timeout
+	}
+}
+
+// OnErrors sets the errors that should be retried.
+func OnErrors(on ...error) RetryOption {
+	return func(rc *RetryConfig) {
+		rc.shouldRetry = func(err error) bool {
+			for _, e := range on {
+				if errors.Is(err, e) {
+					return true
+				}
+			}
+			return false
 		}
 	}
 }
 
-func Poll[T any](pctx context.Context, timeout time.Duration, fn func() (*T, *Err)) (*T, error) {
-	ctx, cancel := context.WithTimeout(pctx, timeout)
+// WithRetryFunc sets the function that determines whether an error should halt the retrier.
+// If the function returns true, the retrier will continue. If it returns false, the retrier will halt.
+func WithRetryFunc(halt func(error) bool) RetryOption {
+	return func(rc *RetryConfig) {
+		rc.shouldRetry = halt
+	}
+}
+
+// Retrier is a struct that can retry an operation until it succeeds or the timeout is reached.
+// The empty struct indicates that the retrier should run for 20 minutes and retry on any non-nil error.
+//
+// Example:
```

Comment:
can you add a compilable example of this, so that it appears in docs properly? https://pkg.go.dev/github.com/databricks/databricks-sdk-go@v0.28.0/service/compute#example-ClusterPoliciesAPI.Create-ClusterPolicies